### ISSUE TITLE
Using Surprise to make predictions on Dataframe or Dataset

### ISSUE THREAD

**Issue by NotaNewbieCoder** [NicolasHug/Surprise#215]
#### Description
I want to know if you can predict ratings from a dataframe or dataobject. I have basically trained a model using your library and can get individual predictions by indicating uid and iid but I need to make predictions for a kaggle competition. Thanks 

#### Steps/Code to Reproduce
#importing all needed and required libraries
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import csv
import surprise
from surprise import Reader, Dataset
from surprise import accuracy
from surprise import SVD

#loading training data and test data
train_data= pd.read_csv(r"C:\Users\Pucado R\Downloads\Data Science Nigeria\train.csv")
test_data= pd.read_csv(r"C:\Users\Pucado R\Downloads\Data Science Nigeria\test.csv")

#data preparation
prepareData= {'itemID': list(train_data.Joke_identifier),
            'userID': list(train_data.Viewers_ID),
            'rating': list(train_data.Rating)}

testdata= {'itemID': list(test_data.Joke_identifier),
            'userID': list(test_data.Viewers_ID),
            'rating': list(test_data.Rating)}

#loading into dataframe
df=pd.DataFrame(prepareData)
dt=pd.DataFrame(testdata)

#reader class using only the rating parameter
reader= Reader(rating_scale=(-5.0, 5.0))

#arranging dataframe
data= Dataset.load_from_df(df[['userID', 'itemID', 'rating']], reader)
testdat= Dataset.load_from_df(dt[['userID', 'itemID', 'rating']], reader)

#build full trainset and use cross validation for evaluation
trainset= data.build_full_trainset()
# Using the SVD algorithm.
algo = SVD()

# Train the algorithm on the trainset
model= algo.fit(trainset)

#predictions using uid and iid
predictions = algo.predict(uid= 'A1', iid= 'Akpororo Lagos 1')
predictions                     #this one works fine but I need to predict on the whole testdat

#predictions using dataset
predictions= algo.predict(testdat)
 

#### Versions
Windows-8.1-6.3.9600-SP0
Python 3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)]
surprise 1.0.6

**Comment by NicolasHug** [comment]
What is keeping you from calling `predict()` on all the samples that you need a prediction for?

**Comment by NotaNewbieCoder** [comment]
There's 43765 samples

**Comment by NicolasHug** [comment]
So? Use a for loop? I really don't get your problem

**Comment by NotaNewbieCoder** [comment]
Thanks you just answered me. For loop works. 😉

**Comment by NotaNewbieCoder** [comment]
Hey man, I'm basically a python newbie and I jut can't figure out how to use the for loop to reiterate over the dataset. Please can you explain?

**Comment by NicolasHug** [comment]
if you have a pandas dataframe something like this should work (you'll have to tweak it):

```python
for _, row in df.iterrrows():
    algo.predict(row.userid, row.itemid)
```

However please note that this is a bug tracker for the surprise package, and those are basic python questions so they would be better addressed on other beginner friendly channels like stack overflow or reddit.

**Comment by poedator** [comment]
Actually it may be faster to call predict on all train samples and then iterate over results. See example provided here:  [https://surprise.readthedocs.io/en/stable/FAQ.html#how-to-get-the-top-n-recommendations-for-each-user](url)

**Comment by ivan-rivera** [comment]
@NicolasHug , first, thank you for developing this library, it's awesome. Just wanted to follow up on your suggestion to generate scores like so:

```
for _, row in df.iterrrows():
    algo.predict(row.userid, row.itemid)
```

Is there a way to predict a batch of user/item IDs? It's just that looping over each row isn't most efficient.

**Comment by gauss1986** [comment]
Actually you can use algo.test(), which works on large number of records. This is what I do:

1. test.loc[:, 'rating'] = 0 # fill in a rating column since it is required in step 2
2. test_processed = Dataset.load_from_df(test[['user_id','book_id','rating']], reader) # use load_from_df to convert the test dataframe to the Dataset format required by step 3
3. NA, test = train_test_split(test_processed, test_size=1.0) # use train_test_split to convert test_processed to list of tuples required in step 4
4. predictions = algo.test(test)
5. est = [i.est for i in predictions] # retrieve the est from prediction results. 

I used this method to predict on 7M records and it takes about 10 mins on my Macbook Pro.

**Comment by nkonyanihlulani** [comment]
@gauss1986 thanks for the approach, it takes less than 1 min on my m1 macbook air to predict 5M records. However, adding an argument on the train_test_split(test_processed, test_size =1.0, shuffle = False) to ensure that the order of the records matches estimates (est)

**Comment by Jean-LucVanZyl** [comment]
@gauss1986 thanks man worked like a charm
